### PR TITLE
Add readme for DataProtection

### PR DESCRIPTION
**Pull request by HaoK**
Part of https://github.com/dotnet/aspnetcore/issues/27561

### CODE REVIEW

**Review by blowdart — APPROVED** [review]

**Comment by HaoK** [comment]
@Rick-Anderson for edits

File: src/DataProtection/README.md [hunk 1]
```diff
@@ -1,8 +1,35 @@
...
+
+## Development Setup
+
+### Build
+
+To build this specific project from source, you can follow the instructions [on building a subset of the code](https://github.com/dotnet/aspnetcore/blob/master/docs/BuildFromSource.md#building-a-subset-of-the-code).
+
+### Test
+
+To run the tests for this project, you can [run the tests on the command line](https://github.com/dotnet/aspnetcore/blob/master/docs/BuildFromSource.md#running-tests-on-command-line) in this directory.
```
Comment (captainsafia) [inline src/DataProtection/README.md:27]:
```suggestion
To run the tests for this project, you can [run the tests on the command line](../../BuildFromSource.md#running-tests-on-command-line) in this directory.
```

File: src/DataProtection/README.md [hunk 2]
```diff
@@ -1,8 +1,35 @@
...
+- `Extensions`: Contains a additional apis via extension methods.
+- `StackExchangeRedis`: Contains the implementation for storing data using StackExchangeREdis
+- `samples`: Contains a collection of sample apps
+- `shared`: Contains a collection of shared constants and helper methods/classes
+
+## Development Setup
+
+### Build
+
+To build this specific project from source, you can follow the instructions [on building a subset of the code](https://github.com/dotnet/aspnetcore/blob/master/docs/BuildFromSource.md#building-a-subset-of-the-code).
```
Comment (captainsafia) [inline src/DataProtection/README.md:23]:
```suggestion
To build this specific project from source, you can follow the instructions [on building a subset of the code](../../BuildFromSource.md#building-a-subset-of-the-code).
```

File: src/DataProtection/README.md [hunk 3]
```diff
@@ -1,8 +1,35 @@
...
+The following contains a description of each sub-directory in the `DataProtection` directory.
+
+- `Abstractions`: Contains the source files for the main DataProtection interfaces like IDataProtector and IDataProtectionProvider
+- `Cryptography.Internal`: Contains the source files for cryptography infrastucture, applications and libraries should not reference this package directly.
+- `Cryptography.KeyDerivation`: Contains the source files related to key derivation, i.e. PBKDF2
+- `DataProtection`: Contains the main implementation of DataProtection for ASP.NET Core to protect and unprotect data.
+- `EntityFrameworkCore`: Contains the implementation for storing data using EntityFrameworkCore
+- `Extensions`: Contains a additional apis via extension methods.
+- `StackExchangeRedis`: Contains the implementation for storing data using StackExchangeREdis
+- `samples`: Contains a collection of sample apps
```
Comment (captainsafia) [inline src/DataProtection/README.md:16]: Are the samples helpful at all in the dev loop?

Comment (HaoK) [reply src/DataProtection/README.md:16]: Probably not, were we not supposed to just document every directory?

Comment (captainsafia) [reply src/DataProtection/README.md:16]: Correct, I was just wondering if there was something helpful outside of running the tests that devs might need to know when making changes in the repo.

Comment (HaoK) [reply src/DataProtection/README.md:16]: Ah gotcha, I mistakenly thought you were implying that we shouldn't mention it...  No these appear to just be customer facing samples/examples for commonly asked scenarios, useful for linking.  I don't believe we build samples, but I think that's true generally, so they are all vulnerable to breaking changes causing samples to no longer work...

File: src/DataProtection/README.md [hunk 4]
```diff
@@ -1,8 +1,35 @@
...
-==============
+# DataProtection
 
 Data Protection APIs for protecting and unprotecting data. You can find documentation for Data Protection in the [ASP.NET Core Documentation](https://docs.microsoft.com/aspnet/core/security/data-protection/).
 
+## Description
+
+The following contains a description of each sub-directory in the `DataProtection` directory.
+
+- `Abstractions`: Contains the source files for the main DataProtection interfaces like IDataProtector and IDataProtectionProvider
```
Comment (Rick-Anderson) [inline src/DataProtection/README.md:9]:
```suggestion
- `Abstractions`: Contains the source files for the main DataProtection interfaces like `IDataProtector` and `IDataProtectionProvider`
```

File: src/DataProtection/README.md [hunk 5]
```diff
@@ -1,8 +1,35 @@
...
+- `Cryptography.Internal`: Contains the source files for cryptography infrastucture, applications and libraries should not reference this package directly.
```
Comment (Rick-Anderson) [inline src/DataProtection/README.md:10]:
```suggestion
- `Cryptography.Internal`: Contains the source files for cryptography infrastucture. Applications and libraries should not reference this package directly.
```

File: src/DataProtection/README.md [hunk 6]
```diff
@@ -1,8 +1,35 @@
...
+- `Cryptography.KeyDerivation`: Contains the source files related to key derivation, i.e. PBKDF2
+- `DataProtection`: Contains the main implementation of DataProtection for ASP.NET Core to protect and unprotect data.
+- `EntityFrameworkCore`: Contains the implementation for storing data using EntityFrameworkCore
+- `Extensions`: Contains additional apis via extension methods.
+- `StackExchangeRedis`: Contains the implementation for storing data using StackExchangeREdis
```
Comment (Rick-Anderson) [inline src/DataProtection/README.md:15]:
```suggestion
- `StackExchangeRedis`: Contains the implementation for storing data using [StackExchange.Redis](https://stackexchange.github.io/StackExchange.Redis/)
```

File: src/DataProtection/README.md [hunk 7]
```diff
@@ -1,8 +1,35 @@
...
+- `samples`: Contains a collection of sample apps
+- `shared`: Contains a collection of shared constants and helper methods/classes
+
+## Development Setup
+
+### Build
+
+To build this specific project from source, you can follow the instructions [on building a subset of the code](../../BuildFromSource.md#building-a-subset-of-the-code).
```
Comment (Rick-Anderson) [inline src/DataProtection/README.md:23]:
```suggestion
To build this specific project from source,  follow the instructions [on building a subset of the code](../../BuildFromSource.md#building-a-subset-of-the-code).
```

File: src/DataProtection/README.md [hunk 8]
```diff
@@ -1,8 +1,35 @@
...
+
+### Test
+
+To run the tests for this project, you can [run the tests on the command line](../../BuildFromSource.md#running-tests-on-command-line) in this directory.
```
Comment (Rick-Anderson) [inline src/DataProtection/README.md:27]:
```suggestion
To run the tests for this project, [run the tests on the command line](../../BuildFromSource.md#running-tests-on-command-line) in this directory.
```

**Review by Rick-Anderson — APPROVED** [review]

**Pull request merged**